x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SLMSnapshotBlockingIntegTests.java

```diff
@@ -231,6 +242,137 @@ public void testRetentionWhileSnapshotInProgress() throws Exception {
         }
     }
 
+    public void testBasicFailureRetention() throws Exception {
+        final String indexName = "test-idx";
+        final String policyId = "test-policy";
+        // Setup
+        logger.info("-->  starting two master nodes and two data nodes");
+        internalCluster().startMasterOnlyNodes(2);
+        internalCluster().startDataOnlyNodes(2);
+
+        createAndPopulateIndex(indexName);
+
+        // Create a snapshot repo
+        initializeRepo(REPO);
+
+        createSnapshotPolicy(policyId, "snap", "1 2 3 4 5 ?", REPO, indexName, true,
+            new SnapshotRetentionConfiguration(null, 1, 2));
+
+        // Create a failed snapshot
+        AtomicReference<String> failedSnapshotName = new AtomicReference<>();
+        {
+            logger.info("-->  stopping random data node, which should cause shards to go missing");
+            internalCluster().stopRandomDataNode();
+            assertBusy(() ->
+                    assertEquals(ClusterHealthStatus.RED, client().admin().cluster().prepareHealth().get().getStatus()),
+                30, TimeUnit.SECONDS);
+
+            final String masterNode = blockMasterFromFinalizingSnapshotOnIndexFile(REPO);
+
+            logger.info("-->  start snapshot");
+            ActionFuture<ExecuteSnapshotLifecycleAction.Response> snapshotFuture = client()
+                .execute(ExecuteSnapshotLifecycleAction.INSTANCE, new ExecuteSnapshotLifecycleAction.Request(policyId));
```

Comment:
There's an `executePolicy` helper that returns the snapshot name as a String (for future tests)